src/hotspot/share/opto/ifnode.cpp

```diff
@@ -455,11 +455,17 @@ static Node* split_if(IfNode *iff, PhaseIterGVN *igvn) {
 }
 
 IfNode* IfNode::make_with_same_profile(IfNode* if_node_profile, Node* ctrl, BoolNode* bol) {
-  if (if_node_profile->Opcode() == Op_If) {
-    return new IfNode(ctrl, bol, if_node_profile->_prob, if_node_profile->_fcnt);
-  } else {
-    assert(if_node_profile->Opcode() == Op_RangeCheck, "only IfNode or RangeCheckNode expected");
+  if (if_node_profile->is_RangeCheck()) {
+    // RangeCheck nodes could be further optimized.
     return new RangeCheckNode(ctrl, bol, if_node_profile->_prob, if_node_profile->_fcnt);
+  } else {
+    // Not a RangeCheckNode? Fall back to IfNode. Assert here that we only try to create a clone from an If node with
+    // the same profiling if that actually makes sense. Some If node subtypes should not be cloned in this way.
+    // In theory, we should not clone BaseCountedLoopEndNodes. But they can end up being used as normal If nodes when
+    // peeling a loop - they serve as zero-trip guard. Allow them as well.
+    assert(if_node_profile->Opcode() == Op_If || if_node_profile->is_RangeCheck()
```

Comment:
`if_node_profile->is_RangeCheck()` seems excluded by the if-else above...
You could move the assert before the if, then it would be correct, I think